### ISSUE TITLE
`F401` false positive with `TypeAliasType` and forward annotation

### ISSUE THREAD

**Issue by Viicos** [astral-sh/ruff#15812]
### Description

search keywords: F401 Typealiastype

```python
from typing import Union

from typing_extensions import TypeAliasType


Json = TypeAliasType(
    'Json',
    'Union[dict[str, Json], list[Json], str, int, float, bool, None]',
)
```

```sh
repro4.py:1:20: F401 [*] `typing.Union` imported but unused
  |
1 | from typing import Union
  |                    ^^^^^ F401
2 |
3 | from typing import TypeAliasType
  |
  = help: Remove unused import: `typing.Union`

```

**Comment by ntBre** [comment]
I'm not a typing expert, but I think this is a true positive. From my understanding, one reason for using forward annotations in general is to avoid imports. Another fix in this case would be to update the second argument to `TypeAliasType` to unquote the `Union` and restrict the quotes to the `Json` parts:

```python
from typing import Union

from typing_extensions import TypeAliasType


Json = TypeAliasType(
    'Json',
    Union[dict[str, 'Json'], list['Json'], str, int, float, bool, None],
)
```

**Comment by Viicos** [comment]
> forward annotations in general is to avoid imports

Indeed, but also in this case to avoid `NameError`s for forward references.

> Another fix in this case would be to update the second argument to `TypeAliasType` to unquote the `Union` and restrict the quotes to the `Json` parts:

While this indeed works, I tend to avoid quoting parts of the annotation, as it doesn't play well with runtime tools. For instance, up until 3.10, a bug is present in `typing.get_type_hints()`:

```python
Test = int

class T:
    t: list['Test']

get_type_hints(T)
#> {'t': list['Test']}  not evaluted!
```

Other issues happened with the deprecated typing aliases (now out of scope because 3.8 is EOL), as for instance `List['Forward']` is implicitly `List[ForwardRef('Forward')]` at runtime, and this caused nasty issues in Pydantic.

**Comment by ntBre** [comment]
Ah okay, thanks for the additional context! I might let someone else (maybe @AlexWaygood) weigh in on whether this is the intended behavior then.

**Comment by AlexWaygood** [comment]
hmm, we probably need to visit the second argument to `TypeAliasType` as a type expression rather than a value expression, similar to what we do with the `bound=` and `default=` keyword arguments to the `TypeVar` constructor (see <https://github.com/astral-sh/ruff/blob/15d886a50213dcd4013db028eeef15a60d182f98/crates/ruff_linter/src/checkers/ast/mod.rs#L1348-L1354>).

@ntBre would you like to take a look? Shouldn't be too hard I don't think, and might be a good way to learn a little more about Ruff's semantic model :-)

**Comment by ntBre** [comment]
Sure, happy to take a look!
